### PR TITLE
Implementation of priority sectors

### PR DESCRIPTION
**Pull request by vegasten**
(Renamed stuff from "highlighting" to "priority" to make the functionality itself less specific.)

Creating of prioritized sectors and adding a table for "treeindex - priority sector" in the hierarchyDB.

**Creation of sectors**
Includes a custom splitter for priority sectors, which is simpler than the usual splitter. It also has an independent budget, making it possible to smaller sectors. 

Budget might need some tweaking, but generally be as small as possible without creating too many sectors. Hence, dependent on how many parts are included in the "prioritized parts". 
    => Including more disciplines, etc. also requires modifying the budget accordingly

**HierarchyDB**
The current solution breaks parallelization by needing to wait for both splitting and hierarchy before adding the new data. Adding the new data is luckily quite quick (~1 second on Troll A), so the total time is not noticably different. It's more a question of code quality and increased complexity..

Note: Some TODOs are still present and in progress.

### CODE REVIEW

File: CadRevealComposer/Operations/StidTagMapper/StidTagMapper.cs [hunk 1]
```diff
@@ -0,0 +1,92 @@
+﻿namespace CadRevealComposer.Operations;
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using Newtonsoft.Json;
+
+public static class StidTagMapper
```
Comment (Strepto) [inline CadRevealComposer/Operations/StidTagMapper/StidTagMapper.cs:9]: Leave the StidTagMapper to another PR, this needs to come in separately

**Review by stigrus — COMMENTED** [review]
A few TODO's left in comments. Not sure if they should be addressed now, removed or kept for future improvements

File: CadRevealComposer/Operations/SectorSplitting/InternalSector.cs [hunk 2]
```diff
@@ -11,5 +11,6 @@ public record InternalSector(
     float MaxNodeDiagonal,
     APrimitive[] Geometries,
     BoundingBox SubtreeBoundingBox,
-    BoundingBox? GeometryBoundingBox
+    BoundingBox? GeometryBoundingBox,
+    bool IsHighlightSector = false
```
Comment (vegasten) [inline CadRevealComposer/Operations/SectorSplitting/InternalSector.cs:15]: Rename to IsPrioritizedSector?

File: CadRevealComposer/Operations/SectorSplitting/PrioritySplittingUtils.cs [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using Primitives;
+using Utils;
+
+public static class PrioritySplittingUtils
+{
+    private static readonly string[] PrioritizedDisciplines = ["PIPE"];
+
+    public static void SetPriorityForHighlightSplittingWithMutation(IReadOnlyList<CadRevealNode> nodes)
```
Comment (vegasten) [inline CadRevealComposer/Operations/SectorSplitting/PrioritySplittingUtils.cs:13]: Rename?

File: CadRevealComposer.Tests/Operations/Splitting/PrioritySplittingUtilsTests.cs [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+using Primitives;
+
+[TestFixture]
+public class PrioritySplittingUtilsTests
+{
+    const string DisciplinePipe = "PIPE";
+    const string DisciplineCable = "CABLE";
+
+    [Test]
+    public void CadRevealNodesWithVariousConfigurations_SetPriorityForHighlightSplittingWithMutation_VerifyCorrectlyMutated()
```
Comment (vegasten) [inline CadRevealComposer.Tests/Operations/Splitting/PrioritySplittingUtilsTests.cs:16]: Rename to priority?

File: CadRevealComposer/Operations/SectorSplitting/PrioritySectorSplitter.cs [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+    )
+    {
+        if (sectorIdGenerator.PeekNextId == 0)
+            _ = sectorIdGenerator.GetNextId(); // Get and discard id 0, as 0 is hardcoded below
+        var dummyRootSector = SplittingUtils.CreateRootSector(0, "/0", new BoundingBox(Vector3.Zero, Vector3.One));
+        yield return dummyRootSector; // Using dummy root sector because we need to remap it to the non-priority sector root later on
+
+        var sectors = new List<InternalSector>();
+        // We split the geometries into sectors based on the discipline because there are very few highlight cases that are cross-discipline
+        // This helps reduce the overhead of a highlight sector, since highlighting of multiple tags within the same discipline is more common
```
Comment (vegasten) [inline CadRevealComposer/Operations/SectorSplitting/PrioritySectorSplitter.cs:25]: Adjust the comment to be more in line with the priority naming?

File: CadRevealComposer/SceneCreator.cs [hunk 6]
```diff
@@ -154,7 +167,10 @@ private static SectorInfo SerializeSector(InternalSector p, string outputDirecto
     {
         var (estimatedTriangleCount, estimatedDrawCalls) = DrawCallEstimator.Estimate(p.Geometries);
 
-        var sectorFilename = p.Geometries.Any() ? $"sector_{p.SectorId}.glb" : null;
+        string? sectorFilename = !p.Geometries.Any()
+            ? null
+            : (!p.IsHighlightSector ? $"sector_{p.SectorId}.glb" : $"highlight_sector_{p.SectorId}.glb"); // The highlight_sector name is used in the frontend to identify the highlight sectors, don't rename without syncing with the frontend.
```
Comment (vegasten) [inline CadRevealComposer/SceneCreator.cs:172]: Call the sectors something else? NOTE: Probably need to change the "guard" in the client (or handle it differently) before doing this

**Review by vegasten — COMMENTED** [review]
Looks good, i think 🤔 just some "highlight" => "prioritized" remaining

Maybe we should take a look a the CadRevealComposerRunner file, to see if some methods/code can be moved to a different file

File: CadRevealComposer/SceneCreator.cs [hunk 7]
```diff
@@ -169,7 +169,7 @@ private static SectorInfo SerializeSector(InternalSector p, string outputDirecto
 
         string? sectorFilename = !p.Geometries.Any()
             ? null
-            : (!p.IsHighlightSector ? $"sector_{p.SectorId}.glb" : $"highlight_sector_{p.SectorId}.glb"); // The highlight_sector name is used in the frontend to identify the highlight sectors, don't rename without syncing with the frontend.
+            : (!p.IsPrioritizedSector ? $"sector_{p.SectorId}.glb" : $"highlight_sector_{p.SectorId}.glb"); // The highlight_sector name is used in the frontend to identify the highlight sectors, don't rename without syncing with the frontend.
```
Comment (stigrus) [inline CadRevealComposer/SceneCreator.cs:172]: Should the filename be updated to `priority_sector_...` as well?

Comment (vegasten) [reply CadRevealComposer/SceneCreator.cs:172]: yes, but we currently have a guard in the client that checks for the "highlight"-prefix, which needs to be changed first. We do have a task for whitelisting instead of blacklisting, which should handle it 🙂

File: CadRevealComposer/Operations/SectorSplitting/PrioritySectorSplitter.cs [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+    )
+    {
+        var nodesProcessed = 0;
+
+        // Sorting by Id as we believe the TreeIndex to group similar parts in the hierarchy together
+        var nodesOrderedByTreeIndex = nodes.OrderBy(x => x.TreeIndex).ToArray();
+
+        while (nodesProcessed < nodes.Length)
+        {
+            // TODO: Should this use spatially aware splitting? Should it place nodes with similar attributes together? Ex: tags?
```
Comment (vegasten) [inline CadRevealComposer/Operations/SectorSplitting/PrioritySectorSplitter.cs:61]: We kinda do this by grouping by discipline and ordering the treeindexes?
Change the TODO to something about grouping by tag, if/when tags get sorted out?

Comment (Strepto) [reply CadRevealComposer/Operations/SectorSplitting/PrioritySectorSplitter.cs:61]: > We kinda do this by grouping by discipline and ordering the treeindexes?
Correct. its very "rough" as we dont split by file or similar, So a sectors extents may be very large if its just inbetween two area files for example, but its better than nothing.

Comment (vegasten) [reply CadRevealComposer/Operations/SectorSplitting/PrioritySectorSplitter.cs:61]: Remove the TODO and create a story regarding "improving grouping of nodes in priority splitting"?

File: CadRevealComposer/Operations/SectorSplitting/PrioritySectorSplitter.cs [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+        // We split the geometries into sectors based on the discipline because there are very few highlight cases that are cross-discipline
+        // This helps reduce the overhead of a priority sector, since highlighting of multiple tags within the same discipline is more common
+        var primitivesGroupedByDiscipline = allGeometries.GroupBy(x => x.Discipline);
+        foreach (var disciplineGroup in primitivesGroupedByDiscipline)
+        {
+            var geometryGroups = disciplineGroup.GroupBy(primitive => primitive.TreeIndex); // Group by treeindex to avoid having one treeindex unnecessary many sectors
+            var nodes = PrioritySplittingUtils.ConvertPrimitiveGroupsToNodes(geometryGroups);
+
+            // Ignore outlier nodes
+            // TODO: Decide if this is the right thing to do
```
Comment (vegasten) [inline CadRevealComposer/Operations/SectorSplitting/PrioritySectorSplitter.cs:33]: Not sure about this one, since there might be exceptions where outlier islands actually are parts of the asset??? In most cases, though, it is reasonable to exclude outliers

File: CadRevealComposer/CadRevealComposerRunner.cs [hunk 10]
```diff
@@ -121,19 +126,33 @@ IReadOnlyList<IModelFormatProvider> modelFormatProviders
...
+        );
 
         if (!exportHierarchyDatabaseTask.IsCompleted)
             Console.WriteLine("Waiting for hierarchy export to complete...");
         exportHierarchyDatabaseTask.Wait();
 
         WriteParametersToParamsFile(modelParameters, composerParameters, outputDirectory);
 
+        // TODO Is this the best place to add sector id mapping to the database?
+        // TODO Cont: Is it ok to mutate the database after it has been exported? Yes?
```
Comment (vegasten) [inline CadRevealComposer/CadRevealComposerRunner.cs:144]: Did we come to a conclusion?

File: CadRevealComposer/CadRevealComposerRunner.cs [hunk 11]
```diff
@@ -84,6 +86,9 @@ IReadOnlyList<IModelFormatProvider> modelFormatProviders
             // collect all nodes for later sector division of the entire scene
             nodesToExport.AddRange(cadRevealNodes);
 
+            // Todo should this return a new list of cadrevealnodes instead of mutating the input?
```
Comment (vegasten) [inline CadRevealComposer/CadRevealComposerRunner.cs:89]: Any reason to not just return a new list?

Comment (stigrus) [reply CadRevealComposer/CadRevealComposerRunner.cs:89]: Maybe to reduce memory usage? Suggest moving a possible refactoring to its own task?

File: CadRevealComposer/Operations/SectorSplitting/PrioritySectorSplitter.cs (hunk 9 above)
Comment (stigrus) [reply CadRevealComposer/Operations/SectorSplitting/PrioritySectorSplitter.cs:33]: Log how many primitives was considered outliers

**Review by Strepto — APPROVED** [review]
🚢
